sirepo/package_data/template/radia/geometry.py.jinja

```diff
@@ -58,35 +57,58 @@ def _add_object(o):
         radia_objs.append(g_id)
     return g_id
 
-id_map = PKDict()
+def _build_radia_geom():
+    id_map = PKDict()
+    radia_objs = []
 {% if isRaw %}
-g_id, id_map = radia_examples.EXAMPLES['{{ exampleName }}']()
+    g_id, id_map = radia_examples.EXAMPLES['{{ exampleName }}']()
 {% elif dmpImportFile %}
-with open('{{ dmpImportFile }}', 'rb') as f:
-    b = f.read()
-    g_id = radia_tk.load_bin(b)
+    with open('{{ dmpImportFile }}', 'rb') as f:
+        b = f.read()
+        g_id = radia_tk.load_bin(b)
 {% else %}
-for obj in sirepo_objs:
-    o = PKDict(obj)
-    if o.model not in _MODELS or o.id in id_map.values():
-        continue
-    o_id = _add_object(o)
-    id_map[o_id] = o.id
-# put all objects in a container for now
-g_id = radia_tk.build_container(radia_objs)
-#id_map[g_id] = '{{ geomId }}'
+    for obj in sirepo_objs:
+        o = PKDict(obj)
+        if o.model not in _MODELS or o.id in id_map.values():
+            continue
+        o_id = _add_object(o, radia_objs, id_map)
+        id_map[o_id] = o.id
+    g_id = radia_tk.build_container(radia_objs)
+{% endif %}
+    # free symmetries to test parallel?
+    #g_id = radia_tk.free_symmetries(g_id)
+    return g_id, id_map
+
+
+def _write_h5_data(data, h5_path):
+    with h5py.File('{{ dataFile }}', 'a') as hf:
+        template_common.dict_to_h5(data, hf, path=h5_path)
+
+
+def _write_dmp(g_id, f_path):
+    with open(f_path, 'wb') as f:
+        f.write(radia_tk.dump_bin(g_id))
+
+
+{% if isParallel %}
+radia_tk.radia_mpi('in')
```

Comment:
This needs to be encapsulated like restrict_op_to_first_rank. I would create a module radia_uti, and have an object which is a [Context Manager](https://docs.python.org/3/library/stdtypes.html#context-manager-types), e.g.

```py
with sirepo.template.radia_uti.MPI() as m:
      m.barrier()
```

The `in` and `off` would be hidden in the context manager, and would be protected against mpi running or not. barrier would be a nop when mpi was not running. That's the reason for the object.

Reply:
a3fd7bd

This works when everything is "correct" but I don't know what to do when they are not.  When I run as though parallel is not available (invoking ```run()``` instead of ```run_background()``` I get 

```Attempting to use an MPI routine before initializing MPICH```

and can't seem to catch it or otherwise guard against it. Do I still need to check isParallel in the jinja template?

I am also unclear as to how I can handle or test the absence of mpi

Reply:
bace751

needed to do this to avoid the init error above

Reply:
8705a6a